File: core/src/main/java/ninja/fido/config/ConfigDataObject.java
/* 
 * Copyright 2017 fido.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.fido.config;

import java.util.Iterator;
import java.util.Map.Entry;
import java.util.Stack;

/**
 *
 * @author David Fiedler
 */
public abstract class ConfigDataObject<T, K, V> implements Iterable<Entry<K, V>> {

	protected final T configObject;

	protected ConfigDataObject parentConfigObject;

	protected Object keyInParent;

	protected String path;

	public String getPath() {
		return path;
	}

	public T getConfigObject() {
		return configObject;
	}

//    public O getConfigObject() {
//		return configObject;
//	}
	public ConfigDataObject(T configObject, ConfigDataObject parentConfigObject, Object keyInParent) {
		this.configObject = configObject;
		this.parentConfigObject = parentConfigObject;
		this.keyInParent = keyInParent;
		createPath();
	}

	public abstract V get(K key);

	public abstract boolean containsKey(K key);

	public abstract void put(K key, V value);
	
	public abstract int getSize();

	public void moveTo(ConfigDataObject parentConfigObject, Object keyInParent) {
		this.parentConfigObject = parentConfigObject;
		this.keyInParent = keyInParent;
		createPath();
	}
	
	public Iterable<ConfigProperty> getIterable() {
		ConfigDataObject configDataObject = this;
		return new Iterable<ConfigProperty>() {

			VariableIterator iterator = new VariableIterator(configDataObject, false);

			@Override
			public Iterator<ConfigProperty> iterator() {
				return iterator;
			}
		};
	}

	public Iterable<ConfigProperty> getVariableIterable() {
		ConfigDataObject configDataObject = this;
		return new Iterable<ConfigProperty>() {

			VariableIterator iterator = new VariableIterator(configDataObject, true);

			@Override
			public Iterator<ConfigProperty> iterator() {
				return iterator;
			}
		};
	}
	
	public String getStringForPrint(){
		StringBuilder out = new StringBuilder();
		for(ConfigProperty configProperty: getIterable()){
			out.append(String.format("%s: %s%n", configProperty.getPath(), configProperty.value));
		}
		return out.toString();
	}
    
    public abstract T getInternalObjects();

	private String createPath() {
		path = "";
		ConfigDataObject currentObject = this;
		while (currentObject.keyInParent != null) {
			Object keyInParentParent = currentObject.keyInParent;
			if (keyInParentParent instanceof String) {
				if(currentObject == this){
					path = (String) keyInParentParent;
				}
				else if(path.startsWith("[")){
					path = keyInParentParent + path;
				}
				else{
					path = keyInParentParent + "." + path;
				}
			}
			else {
				path = "[" + Integer.toString((int) keyInParentParent) + "]." + path;
			}
			currentObject = currentObject.parentConfigObject;
		}
		return path;
	}

	class VariableIterator implements Iterator<ConfigProperty> {

		private final Stack<VariableIteratorContext> contextStack;
		
		private final boolean varsOnly;

		private VariableIteratorContext currentContext;

		private Iterator<Entry<K, V>> currentIterator;

		private Entry<K, V> currentEntry;

		VariableIterator(ConfigDataObject configDataObject, boolean varsOnly) {
			this.varsOnly = varsOnly;
			contextStack = new Stack<>();
			currentContext = new VariableIteratorContext(configDataObject, configDataObject.iterator());
			currentIterator = currentContext.iterator;
		}

		@Override
		public boolean hasNext() {
			if (currentEntry != null && isRequestedType(currentEntry.getValue())) {
				return true;
			}
			checkIterator();
			while (currentIterator.hasNext()) {

				/* fetching next entry */
				currentEntry = currentIterator.next();
				Object currentValue = currentEntry.getValue();

				if (isRequestedType(currentValue)) {
					return true;
				}
				else if (currentValue instanceof ConfigDataObject) {
					contextStack.push(currentContext);
					ConfigDataObject currentConfigObject = (ConfigDataObject) currentValue;
					currentContext = new VariableIteratorContext(currentConfigObject, currentConfigObject.iterator());
					currentIterator = currentContext.iterator;
				}
				checkIterator();
			}
			return false;
		}

		@Override
		public ConfigProperty next() {
			if (hasNext()) {
				ConfigProperty property = new ConfigProperty(
						currentContext.configDataObject, currentEntry.getKey(), currentEntry.getValue());
				currentEntry = null;
				return property;
			}
			else {
				return null;
			}
		}
		
		private boolean isRequestedType(Object value){
			if(varsOnly){
				return Parser.containsVariable(value);
			}
			else{
				return !(value instanceof ConfigDataObject);
			}
		}

		private void checkIterator() {
			while (!currentIterator.hasNext() && !contextStack.empty()) {
				currentContext = contextStack.pop();
				currentIterator = currentContext.iterator;
			}
		}

		private class VariableIteratorContext {

			private final ConfigDataObject configDataObject;

			private final Iterator<Entry<K, V>> iterator;

			VariableIteratorContext(ConfigDataObject configDataObject, Iterator<Entry<K, V>> iterator) {
				this.configDataObject = configDataObject;
				this.iterator = iterator;
			}
		}

	}
}

File: python/config_data_object.py

from typing import Callable
from config_property import ConfigProperty


class ConfigDataObject:

	def __init__(self, parent_config_object, key_in_parent, is_array=False):
		self.is_array = is_array
		self.config_object = {}
		self.parent_config_object = parent_config_object
		self.key_in_parent = key_in_parent
		self.path = self.create_path()

	def __iter__(self):
		return iter(self.config_object)

	def items(self):
		return self.config_object.items()

	def put(self, key, value):
		self.config_object[key] = value

	def get_size(self):
		return len(self.config_object)

	def move_to(self, parent_config_object, key_in_parent):
		self.parent_config_object = parent_config_object
		self.key_in_parent = key_in_parent
		self.create_path()

	def get_string_for_print(self):
		out = ""

		def append_line(config_property, out):
			out += "{}: {}%n".format(config_property.get_path(), config_property.value)

		self.iterate_properties(lambda x: True, append_line)

	def create_path(self):
		path = ""
		current_object = self
		while current_object.key_in_parent:
			key_in_parent = current_object.keyInParent
			if isinstance(key_in_parent, int):
				path = "[{}].{}".format(key_in_parent, path)
			else:
				if current_object == self:
					path = key_in_parent
				elif path.startsWith("["):
					path = key_in_parent + path
				else:
					path = key_in_parent + "." + path
			current_object = current_object.parent_config_object
		return path

	def iterate_properties(
			self, filter_function: Callable[any, bool], iter_function: Callable[ConfigProperty, any, None], out=None):
		for key, value in self.items():
			if isinstance(value, ConfigDataObject):
				value.iterate_properties(filter_function, iter_function)
			elif filter_function(value):
				config_property = ConfigProperty(self, key, value)
				iter_function(config_property, out)

	# class Entry:
	#     def __init__(self, key, value):
	#         self.key = key
	#         self.value = value
	#
	# class VariableIterator:
	#
	#     def __init__(self, self, vars_only):
	#         self.context_stack = []
	#         self.vars_only = vars_only
	#         self.current_context = self.VariableIteratorContext(self, 0)
	#         self.current_object = self;
	#         self.current_entry = None;
	#
	#     def __iter__(self):
	#         return self
	#
	#     def next(self):
	#         # if not self.current_entry is None and self.is_requested_type(self.current_entry.value))
	#         # self.check_iterator()
	#
	#         while not self.current_context.finished() or self.context_stack:
	#             if self.current_context.finished():
	#                 self.current_context = self.context_stack.pop()
	#                 self.current_object = self.current_context.self
	#             else:
	#                 self.current_entry =
	#                 current_value = self.current_entry.value
	#                 if self.is_requested_type(current_value):
	#                     return ConfigProperty(self.current_object, self.current_entry.key, current_value)
	#                 elif isinstance(current_value, ConfigDataObject):
	#                     self.context_stack.push(self.current_context)
	#                     self.current_contex = self.VariableIteratorContext(current_value, 0)
	#                 self.check_iterator()
	#
	#     def is_requested_type(self, value):
	#         if self.varsOnly:
	#             return parser.contains_variable(value)
	#         else:
	#             return not isinstance(value, ConfigDataObject)
	#
	#     def check_iterator(self):
	#         while self.current_context.finished() and self.context_stack:
	#             self.current_context = self.context_stack.pop()
	#             self.current_object = self.current_context.self
	#
	#
	#     class VariableIteratorContext:
	#
	#         def __init__(self, self, position):
	#             self.self = self
	#             self.position = position
	#
	#         def finished(self):
	#             return self.position >= len(self.self)
	#
	#         def increment(self):
	#             self.position += 1